src/factory-reset/factory-reset-tool.c

```diff
@@ -0,0 +1,374 @@
+                (void) retrigger_block_devices();
+
+        return 0;
+}
+
+static int vl_method_get_factory_reset_mode(sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
+        int r;
+
+        assert(parameters);
+
+        r = sd_varlink_dispatch(link, parameters, /* table= */ NULL, /* userdata= */ NULL);
+        if (r != 0)
+                return r;
+
+        FactoryResetMode f = factory_reset_mode();
+        if (f < 0)
+                return f;
+
+        return sd_varlink_replybo(link, SD_JSON_BUILD_PAIR_STRING("mode", factory_reset_mode_to_string(f)));
+}
+
+static int vl_method_can_request_factory_reset(sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
+        int r;
+
+        assert(parameters);
+
+        r = sd_varlink_dispatch(link, parameters, /* table= */ NULL, /* userdata= */ NULL);
+        if (r != 0)
+                return r;
+
+        return sd_varlink_replybo(link, SD_JSON_BUILD_PAIR_BOOLEAN("supported", is_efi_boot()));
+}
+
+static int varlink_service(void) {
+        int r;
+
+        /* Invocation as Varlink service */
+
+        _cleanup_(sd_varlink_server_unrefp) sd_varlink_server *varlink_server = NULL;
+        r = varlink_server_new(&varlink_server, SD_VARLINK_SERVER_ROOT_ONLY, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to allocate Varlink server: %m");
+
+        r = sd_varlink_server_add_interface(varlink_server, &vl_interface_io_systemd_FactoryReset);
+        if (r < 0)
+                return log_error_errno(r, "Failed to add Varlink interface: %m");
+
+        r = sd_varlink_server_bind_method_many(
+                        varlink_server,
+                        "io.systemd.FactoryReset.GetFactoryResetMode",    vl_method_get_factory_reset_mode,
+                        "io.systemd.FactoryReset.CanRequestFactoryReset", vl_method_can_request_factory_reset);
+        if (r < 0)
+                return log_error_errno(r, "Failed to bind Varlink methods: %m");
+
+        r = sd_varlink_server_loop_auto(varlink_server);
+        if (r < 0)
+                return log_error_errno(r, "Failed to run Varlink event loop: %m");
+
+        return EXIT_SUCCESS;
+}
+
+static int run(int argc, char *argv[]) {
+        static const Verb verbs[] = {
+                { "status",   VERB_ANY, 1, VERB_DEFAULT, verb_status   },
+                { "request",  VERB_ANY, 1, 0,            verb_request  },
+                { "cancel",   VERB_ANY, 1, 0,            verb_cancel   },
+                { "complete", VERB_ANY, 1, 0,            verb_complete },
+                {}
+        };
+
+        int r;
+
+        log_setup();
+
+        r = parse_argv(argc, argv);
+        if (r <= 0)
+                return r;
+
+        if (arg_varlink)
+                return varlink_service();
+
+        return dispatch_verb(argc, argv, verbs, /* userdata= */ NULL);
+}
+
+DEFINE_MAIN_FUNCTION(run);
```

Comment:
`DEFINE_MAIN_FUNCTION_WITH_POSITIVE_FAILURE`
Otherwise, the exit code from `verb_status()` is not propagated.